src/drivers/intel/cavs/ipc.c

```diff
@@ -185,6 +194,15 @@ enum task_state ipc_platform_do_cmd(struct ipc *ipc)
 	hdr = mailbox_validate();
 #endif
 
+#ifdef IPC4_EARLY_ACK
+	/*
+	 * notify host that we have received the message and
+	 * do this before we send the response
+	 */
+	ipc_write(IPC_DIPCTDR, ipc_read(IPC_DIPCTDR) | IPC_DIPCTDR_BUSY);
+	ipc_write(IPC_DIPCTDA, ipc_read(IPC_DIPCTDA) | IPC_DIPCTDA_DONE);
+#endif
+
 	/* perform command */
 	ipc_cmd(hdr);
```

Comment:
Right, I see. the reply is sent as a new message within `ipc_cmd()` at the end of the handling.
In theory host software could rely on the BUSY to decide to send a new message or not, so if we clear it before we processed the message from the registers and mailbox then there is a chance that the host changes the content under our feet...

One way to be safer is to add a flag to `struct ipc_msg`, something like `bool reply_message`. If it is true then do the ack before sending the reply message itself. The flag is ignored for IPC3.